components/sup/src/event/nitox.rs

```diff
@@ -0,0 +1,90 @@
+                              let EventConnectionInfo { name,
+                                                        verbose,
+                                                        cluster_uri,
+                                                        cluster_id,
+                                                        auth_token, } = conn_info;
+                              let cc =
+                                  ConnectCommand::builder().name(Some(name))
+                                                           .verbose(verbose)
+                                                           .auth_token(Some(auth_token.to_string()))
+                                                           .tls_required(false)
+                                                           .build()
+                                                           .expect("Could not create NATS \
+                                                                    ConnectCommand");
+                              let opts =
+                                  NatsClientOptions::builder().connect_command(cc)
+                                                              .cluster_uri(cluster_uri.as_str())
+                                                              .build()
+                                                              .expect("Could not create \
+                                                                       NatsClientOptions");
+
+                              let publisher = NatsClient::from_options(opts).map_err(|e| {
+                                                  error!("Error connecting to NATS: {}", e);
+                                                  e.into()
+                                              })
+                                              .and_then(|client| {
+                                                  NatsStreamingClient::from(client)
+                        .cluster_id(cluster_id)
+                        .connect()
+                                              })
+                                              .map_err(|streaming_error| {
+                                                  error!("Error upgrading to streaming NATS \
+                                                          client: {}",
+                                                         streaming_error)
+                                              })
+                                              .and_then(move |client| {
+                                                  sync_tx.send(()).expect("Couldn't synchronize \
+                                                                           event thread!");
+                                                  event_rx.for_each(move |event: Vec<u8>| {
+                                                              let publish_event = client
+                            .publish(HABITAT_SUBJECT.into(), event.into())
+                            .map_err(|e| {
```

Comment:
Is this doing a necessary mapping, or just in here to print the error?

If the latter, perhaps use `inspect`?

Reply:
It's doing both, actually.

In order to spawn a future on an executor, both its `Item` and `Error` type must be `()`, because once you spawn it, it's off on a threadpool somewhere and can't really return anything to you. You can combine futures however you want to, but in the end, you must map both types to `()`.

(If you turn your head and squint, a spawnable future is a bit like a function that returns `std::result::Result<(),()>`.)

Thus, this `map_err` is taking any error in publishing an event and surfacing it to the logs with `error!(...)`. Since this also has a value of `()`, it also serves to get the future into a state where we can spawn it. (The `publish` future's `Item` type is already `()`, so we don't need an accompanying `map` call.)

I thought that changing it to 

```rust
...
.map_err(|e| {
  error!(...);
  ()
});
```
would trigger Clippy, but it doesn't appear to. However, while more explicit, that strikes me as needlessly noisy.